### PR TITLE
PyTorch Global Magnitude Pruning

### PR DESCRIPTION
**Pull request by bfineran**
enable global magnitude pruning by adding `global_sparsity: True` to GMPruningModifier configs

* [x] refactor mask creators and masking objects to process lists of tensors instead of single tensors
* [x] add option for global masking in mask creators and pruning modifier
* [x] testing

### CODE REVIEW

**Comment by bfineran** [comment]
Current status is this PR will be on hold until other algorithms are landed on main

**Comment by bfineran** [comment]
Putting this back up for review

**Review by kevinaer — DISMISSED** [review]

**Review by markurtz — DISMISSED** [review]

**Comment by markurtz** [comment]
Overall looks good, 2 notes for the future to keep an eye on:

- I'm worried about the memory hit we're going to take on creating a new tensor the size of all the params that are used to just calculate the threshold. This may be something we need to do on the CPU, or at least have an option for, instead of keeping the tensor on the current device
- Also worried about the discoverability of global magnitude. As we're doing the refactor and renaming, it may make sense to break out GradualMagnitude and GlobalMagnitude into their own root level classes that extend the same base class so it's more discoverable and readily available in docs and tutorials

**Review by natuan — DISMISSED** [review]
LGTM. For future iteration, I wonder if the `create_sparsity_masks` could be simplified by always create mask with the defined sparsity on top of all tensors passed in. The distinction between global and layer-wise would be made from the call-sites by either passing a tensor list, or calling this func on separate tensors (there's a burden for layer wise here, but might worth it to simplify the create mask code)

**Comment by bfineran** [comment]
@markurtz updated to include child classes for pruning modifiers, note that we will have to be careful about `ModifierProp` serialization when inheriting from a class with more props.  Agreed on the memory issue, definitely something to keep an eye out for.

@natuan That implementation would move logic for when to group globally or not to two different places (the call site would have to pass in tensors separately), as implemented, the only place in the stack that cares about global or not, other than propagating, is the mask creator.  I do think adding support for a single tensor input would be good though.

**Review by markurtz — DISMISSED** [review]

File: tests/sparseml/pytorch/optim/test_modifier_pruning.py
Comment (kevinaer) [inline tests/sparseml/pytorch/optim/test_modifier_pruning.py:397]: Can you remove the skips? I am not sure if it is being used in some legacy job but environment based skipping is no longer being used in favor of make targets.

File: tests/sparseml/pytorch/optim/test_modifier_pruning.py
Comment (kevinaer) [inline tests/sparseml/pytorch/optim/test_modifier_pruning.py:477]: Can you remove the skips? I am not sure if it is being used in some legacy job but environment based skipping is no longer being used in favor of make targets.

**Review by kevinaer — CHANGES_REQUESTED** [review]

**Comment by bfineran** [comment]
@kevinaer resolved, thanks!

**Review by kevinaer — APPROVED** [review]

**Review by markurtz — APPROVED** [review]

**Pull request merged**
